<b>Machine Learning Tutorial - Naive Bayes: Exercise</b>
</br>
Use wine dataset from sklearn.datasets to classify wines into 3 categories. Load the dataset and split it into test and train. After that train the model using Gaussian and Multinominal classifier and post which model performs better. Use the trained model to perform some predictions on test data.

In [80]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_wine

In [81]:
wine=load_wine()
dir(wine)

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']

In [82]:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [83]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [84]:
df = pd.DataFrame(wine.data, columns=wine.feature_names)
df['target']=wine.target
df['target_name']=df['target'].apply(lambda x: wine.target_names[x])
df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  target target_name  
0                            3.92   1065.0       0     class_0  
1                            3.40   1050.0       0     class_0  
2                            3.17   1185.0       0     class_0  
3                            3.45   1480.0       0     class_0  
4                            2.93    735.0       0     class_0  
..                            ...      ...     ...         ...  
173                          1.74    740.0       2     class_2  
174                          1.56    750.0       2     class_2  
175                          1.56    835.0       2     class_2  
176                          1.62    840.0       2     class_2  
177                          1.60    560.0       2     class_2  

[178 rows x 15 columns]

In [85]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test=train_test_split(wine.data, wine.target, test_size=0.2, random_state=42)

In [86]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()

In [87]:
model.fit(x_train,y_train)

GaussianNB()

In [88]:
model.score(x_test,y_test)

1.0

In [89]:
y_test

array([0, 0, 2, 0, 1, 0, 1, 2, 1, 2, 0, 2, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 2, 2, 2, 1, 1, 1, 0, 0, 1, 2, 0, 0, 0])

In [90]:
y_pred = model.predict(x_test)
y_pred

array([0, 0, 2, 0, 1, 0, 1, 2, 1, 2, 0, 2, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 2, 2, 2, 1, 1, 1, 0, 0, 1, 2, 0, 0, 0])

In [91]:
from sklearn.model_selection import cross_val_score

In [92]:
acc=0
n=0
for i in range(2,26):
    arr = cross_val_score(GaussianNB(), wine.data, wine.target, cv=i)
    curr_a = 100*np.average(arr)
    if(curr_a>acc):
        acc=curr_a
        n=i
    print("cv: {}, accuracy: {}".format(i, curr_a))
print("max accuracy is for cv: {}, accuracy: {}".format(n, acc))

cv: 2, accuracy: 97.75280898876404
cv: 3, accuracy: 96.07344632768361
cv: 4, accuracy: 96.09848484848484
cv: 5, accuracy: 96.63492063492063
cv: 6, accuracy: 96.0919540229885
cv: 7, accuracy: 96.13186813186813
cv: 8, accuracy: 96.1215415019763
cv: 9, accuracy: 95.55555555555557
cv: 10, accuracy: 97.77777777777779
cv: 11, accuracy: 97.22593582887701
cv: 12, accuracy: 97.77777777777779
cv: 13, accuracy: 96.66103127641588
cv: 14, accuracy: 96.65750915750915
cv: 15, accuracy: 96.66666666666666
cv: 16, accuracy: 97.77462121212122
cv: 17, accuracy: 97.2727272727273
cv: 18, accuracy: 97.22222222222221
cv: 19, accuracy: 97.77777777777776
cv: 20, accuracy: 96.66666666666666
cv: 21, accuracy: 96.75925925925925
cv: 22, accuracy: 97.22222222222221
cv: 23, accuracy: 97.28260869565217
cv: 24, accuracy: 97.32142857142857
cv: 25, accuracy: 97.21428571428572
max accuracy is for cv: 10, accuracy: 97.77777777777779


Using Multinomial Naive Bayes

In [93]:
from sklearn.naive_bayes import MultinomialNB
mmodel=MultinomialNB()
mmodel.fit(x_train, y_train)

MultinomialNB()

In [94]:
model.score(x_test,y_test)

1.0

In [96]:
acc=0
n=0
for i in range(2,26):
    arr = cross_val_score(MultinomialNB(), wine.data, wine.target, cv=i)
    curr_a = 100*np.average(arr)
    if(curr_a>acc):
        acc=curr_a
        n=i
    print("cv: {}, accuracy: {}".format(i, curr_a))
print("max accuracy is for cv: {}, accuracy: {}".format(n, acc))

cv: 2, accuracy: 83.14606741573034
cv: 3, accuracy: 83.21092278719398
cv: 4, accuracy: 83.27020202020202
cv: 5, accuracy: 84.3968253968254
cv: 6, accuracy: 84.38697318007662
cv: 7, accuracy: 84.43956043956044
cv: 8, accuracy: 86.04249011857708
cv: 9, accuracy: 86.11111111111111
cv: 10, accuracy: 84.9673202614379
cv: 11, accuracy: 86.02941176470587
cv: 12, accuracy: 84.40476190476191
cv: 13, accuracy: 84.53085376162299
cv: 14, accuracy: 85.62271062271061
cv: 15, accuracy: 85.5050505050505
cv: 16, accuracy: 84.94318181818181
cv: 17, accuracy: 85.77540106951874
cv: 18, accuracy: 85.49382716049382
cv: 19, accuracy: 85.14619883040936
cv: 20, accuracy: 85.48611111111111
cv: 21, accuracy: 85.3174603174603
cv: 22, accuracy: 84.91161616161617
cv: 23, accuracy: 86.33540372670808
cv: 24, accuracy: 86.23511904761904
cv: 25, accuracy: 85.42857142857143
max accuracy is for cv: 23, accuracy: 86.33540372670808


The Gaussian Naive Bayes Classifier is giving a significant higher accuracy than the Multinomial Naive Bayes</br>
The accuracy of Naive Bayes for a 10-Fold cross validation is around 97.78% </br>
The accuracy of Naive Bayes for a 23-Fold cross validation is around 86.34% 